### ISSUE TITLE
libgit2sharp 0.26.0-preview-0062 is failing on Ubuntu

### ISSUE THREAD

**Issue by AnthonySteele** [NuKeeperDotNet/NuKeeper#539]
libgit2sharp is failing, again.

On Ubuntu today:

```
Failed on repo AwsWatchman TypeInitializationException : The type initializer for 'LibGit2Sharp.Core.NativeMethods' threw an exception.
   at LibGit2Sharp.Core.NativeMethods.git_clone(git_repository*& repo, String origin_url, FilePath workdir_path, GitCloneOptions& opts)
   at LibGit2Sharp.Core.Proxy.git_clone(String url, String workdir, GitCloneOptions& opts)
   at LibGit2Sharp.Repository.Clone(String sourceUrl, String workdirPath, CloneOptions options)
   at NuKeeper.Git.LibGit2SharpDriver.Clone(Uri pullEndpoint) in C:\Code\NuKeeper\NuKeeper\Git\LibGit2SharpDriver.cs:line 43
   at NuKeeper.Engine.RepositoryUpdater.GitInit(IGitDriver git, RepositoryData repository) in C:\Code\NuKeeper\NuKeeper\Engine\RepositoryUpdater.cs:line 108
   at NuKeeper.Engine.RepositoryUpdater.Run(IGitDriver git, RepositoryData repository, SettingsContainer settings) in C:\Code\NuKeeper\NuKeeper\Engine\RepositoryUpdater.cs:line 49
   at NuKeeper.Engine.GitRepositoryEngine.Run(RepositorySettings repository, UsernamePasswordCredentials gitCreds, Identity userIdentity, SettingsContainer settings) in C:\Code\NuKeeper\NuKeeper\Engine\GitRepositoryEngine.cs:line 56

Segmentation fault (core dumped)
```

This is either an issue with [LibGit2Sharp version `0.26.0-preview-0062`](https://www.nuget.org/packages/LibGit2Sharp/0.26.0-preview-0062), its dependencies, or OS version `Ubuntu 18.04.1 LTS` or the interaction of the two.

I have verified that `libcurl3` is installed, as `LibGit2Sharp` needs this.

**Comment by slang25** [comment]
I think we might need to wait for this to fix these issues:
https://github.com/libgit2/libgit2sharp/pull/1618

**Comment by AnthonySteele** [comment]
I've seen a similar issue a couple of months ago; that was solved by installing `libcurl3` [as mentioned here](https://github.com/terrajobst/git-istage/issues/13#issuecomment-430895458). 

But today, that fix doesn't appear to be working.

**Comment by ethomson** [comment]
Oy, sorry for this.  Can you post the output of a couple of commands?

`uname -a`
`cat /etc/debian_version`

Can you find the libgit2 native binary (included in the LibGit2Sharp.NativeBinaries package) and paste the output of `ldd libgit2-8e0b172.so`?

There should be several `libgit2-8e0b172.so` files, I'm particularly interested in the one in `runtimes/ubuntu.18.04-x64/native` and `runtimes/linux-x64/native`...

**Comment by AnthonySteele** [comment]
```
Welcome to Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-1027-aws x86_64)

$ uname -a

linux ip-172-x-x-x 4.15.0-1027-aws #27-Ubuntu SMP Fri Nov 2 15:14:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$  cat /etc/debian_version

buster/sid

$ cd ~/.dotnet/tools/.store/nukeeper/0.10.0/nukeeper/0.10.0/tools/netcoreapp2.1/any/runtimes/ubuntu.18.04-x64/native

$ ldd libgit2-8e0b172.so

./libgit2-8e0b172.so: /usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_4' not found (required by ./libgit2-8e0b172.so)
        linux-vdso.so.1 (0x00007fff587da000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f2dd4f57000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2dd4d38000)
        libcurl.so.4 => /usr/lib/x86_64-linux-gnu/libcurl.so.4 (0x00007f2dd4ab9000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f2dd484f000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f2dd43d7000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2dd41ba000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2dd3dc9000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2dd547b000)
        libnghttp2.so.14 => /usr/lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f2dd3ba4000)
        libidn2.so.0 => /usr/lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f2dd3987000)
        librtmp.so.1 => /usr/lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f2dd376b000)
        libpsl.so.5 => /usr/lib/x86_64-linux-gnu/libpsl.so.5 (0x00007f2dd355d000)
        libssl.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f2dd32f5000)
        libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f2dd2eb2000)
        libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f2dd2c67000)
        libldap_r-2.4.so.2 => /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 (0x00007f2dd2a15000)
        liblber-2.4.so.2 => /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 (0x00007f2dd2807000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f2dd2603000)
        libunistring.so.2 => /usr/lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f2dd2285000)
        libgnutls.so.30 => /usr/lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f2dd1f20000)
        libhogweed.so.4 => /usr/lib/x86_64-linux-gnu/libhogweed.so.4 (0x00007f2dd1cec000)
        libnettle.so.6 => /usr/lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f2dd1ab6000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f2dd1835000)
        libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f2dd155f000)
        libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f2dd132d000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f2dd1129000)
        libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f2dd0f1e000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f2dd0d03000)
        libsasl2.so.2 => /usr/lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f2dd0ae8000)
        libgssapi.so.3 => /usr/lib/x86_64-linux-gnu/libgssapi.so.3 (0x00007f2dd08a7000)
        libp11-kit.so.0 => /usr/lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f2dd0578000)
        libtasn1.so.6 => /usr/lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f2dd0365000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f2dd0161000)
        libheimntlm.so.0 => /usr/lib/x86_64-linux-gnu/libheimntlm.so.0 (0x00007f2dcff58000)
        libkrb5.so.26 => /usr/lib/x86_64-linux-gnu/libkrb5.so.26 (0x00007f2dcfccb000)
        libasn1.so.8 => /usr/lib/x86_64-linux-gnu/libasn1.so.8 (0x00007f2dcfa29000)
        libhcrypto.so.4 => /usr/lib/x86_64-linux-gnu/libhcrypto.so.4 (0x00007f2dcf7f3000)
        libroken.so.18 => /usr/lib/x86_64-linux-gnu/libroken.so.18 (0x00007f2dcf5dd000)
        libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f2dcf3d5000)
        libwind.so.0 => /usr/lib/x86_64-linux-gnu/libwind.so.0 (0x00007f2dcf1ac000)
        libheimbase.so.1 => /usr/lib/x86_64-linux-gnu/libheimbase.so.1 (0x00007f2dcef9d000)
        libhx509.so.5 => /usr/lib/x86_64-linux-gnu/libhx509.so.5 (0x00007f2dced53000)
        libsqlite3.so.0 => /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f2dcea4a000)
        libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f2dce812000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2dce474000)
		
		
Welcome to Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-1027-aws x86_64)

$ uname -a

linux ip-172-x-x-x 4.15.0-1027-aws #27-Ubuntu SMP Fri Nov 2 15:14:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$  cat /etc/debian_version

buster/sid

$ cd ~/.dotnet/tools/.store/nukeeper/0.10.0/nukeeper/0.10.0/tools/netcoreapp2.1/any/runtimes/ubuntu.18.04-x64/native

$ ldd libgit2-8e0b172.so

./libgit2-8e0b172.so: /usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_4' not found (required by ./libgit2-8e0b172.so)
        linux-vdso.so.1 (0x00007fff587da000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f2dd4f57000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2dd4d38000)
        libcurl.so.4 => /usr/lib/x86_64-linux-gnu/libcurl.so.4 (0x00007f2dd4ab9000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f2dd484f000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f2dd43d7000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2dd41ba000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2dd3dc9000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2dd547b000)
        libnghttp2.so.14 => /usr/lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f2dd3ba4000)
        libidn2.so.0 => /usr/lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f2dd3987000)
        librtmp.so.1 => /usr/lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f2dd376b000)
        libpsl.so.5 => /usr/lib/x86_64-linux-gnu/libpsl.so.5 (0x00007f2dd355d000)
        libssl.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f2dd32f5000)
        libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f2dd2eb2000)
        libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f2dd2c67000)
        libldap_r-2.4.so.2 => /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 (0x00007f2dd2a15000)
        liblber-2.4.so.2 => /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 (0x00007f2dd2807000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f2dd2603000)
        libunistring.so.2 => /usr/lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f2dd2285000)
        libgnutls.so.30 => /usr/lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f2dd1f20000)
        libhogweed.so.4 => /usr/lib/x86_64-linux-gnu/libhogweed.so.4 (0x00007f2dd1cec000)
        libnettle.so.6 => /usr/lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f2dd1ab6000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f2dd1835000)
        libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f2dd155f000)
        libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f2dd132d000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f2dd1129000)
        libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f2dd0f1e000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f2dd0d03000)
        libsasl2.so.2 => /usr/lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f2dd0ae8000)
        libgssapi.so.3 => /usr/lib/x86_64-linux-gnu/libgssapi.so.3 (0x00007f2dd08a7000)
        libp11-kit.so.0 => /usr/lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f2dd0578000)
        libtasn1.so.6 => /usr/lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f2dd0365000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f2dd0161000)
        libheimntlm.so.0 => /usr/lib/x86_64-linux-gnu/libheimntlm.so.0 (0x00007f2dcff58000)
        libkrb5.so.26 => /usr/lib/x86_64-linux-gnu/libkrb5.so.26 (0x00007f2dcfccb000)
        libasn1.so.8 => /usr/lib/x86_64-linux-gnu/libasn1.so.8 (0x00007f2dcfa29000)
        libhcrypto.so.4 => /usr/lib/x86_64-linux-gnu/libhcrypto.so.4 (0x00007f2dcf7f3000)
        libroken.so.18 => /usr/lib/x86_64-linux-gnu/libroken.so.18 (0x00007f2dcf5dd000)
        libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f2dcf3d5000)
        libwind.so.0 => /usr/lib/x86_64-linux-gnu/libwind.so.0 (0x00007f2dcf1ac000)
        libheimbase.so.1 => /usr/lib/x86_64-linux-gnu/libheimbase.so.1 (0x00007f2dcef9d000)
        libhx509.so.5 => /usr/lib/x86_64-linux-gnu/libhx509.so.5 (0x00007f2dced53000)
        libsqlite3.so.0 => /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f2dcea4a000)
        libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f2dce812000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2dce474000)
		
```

**Comment by AnthonySteele** [comment]
And:

```
 cd ~/.dotnet/tools/.store/nukeeper/0.10.0/nukeeper/0.10.0/tools/netcoreapp2.1/any/runtimes/linux-x64/native

 ldd libgit2-8e0b172.so
        linux-vdso.so.1 (0x00007ffc0adfd000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f75a88d4000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f75a86b5000)
        libcurl.so.4 => /usr/lib/x86_64-linux-gnu/libcurl.so.4 (0x00007f75a8436000)
        libssl.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f75a81ce000)
        libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f75a7d8b000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f75a7b6e000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f75a777d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f75a8dee000)
        libnghttp2.so.14 => /usr/lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f75a7558000)
        libidn2.so.0 => /usr/lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f75a733b000)
        librtmp.so.1 => /usr/lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f75a711f000)
        libpsl.so.5 => /usr/lib/x86_64-linux-gnu/libpsl.so.5 (0x00007f75a6f11000)
        libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f75a6cc6000)
        libldap_r-2.4.so.2 => /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 (0x00007f75a6a74000)
        liblber-2.4.so.2 => /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 (0x00007f75a6866000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f75a6662000)
        libunistring.so.2 => /usr/lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f75a62e4000)
        libgnutls.so.30 => /usr/lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f75a5f7f000)
        libhogweed.so.4 => /usr/lib/x86_64-linux-gnu/libhogweed.so.4 (0x00007f75a5d4b000)
        libnettle.so.6 => /usr/lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f75a5b15000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f75a5894000)
        libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f75a55be000)
        libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f75a538c000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f75a5188000)
        libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f75a4f7d000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f75a4d62000)
        libsasl2.so.2 => /usr/lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f75a4b47000)
        libgssapi.so.3 => /usr/lib/x86_64-linux-gnu/libgssapi.so.3 (0x00007f75a4906000)
        libp11-kit.so.0 => /usr/lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f75a45d7000)
        libtasn1.so.6 => /usr/lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f75a43c4000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f75a41c0000)
        libheimntlm.so.0 => /usr/lib/x86_64-linux-gnu/libheimntlm.so.0 (0x00007f75a3fb7000)
        libkrb5.so.26 => /usr/lib/x86_64-linux-gnu/libkrb5.so.26 (0x00007f75a3d2a000)
        libasn1.so.8 => /usr/lib/x86_64-linux-gnu/libasn1.so.8 (0x00007f75a3a88000)
        libhcrypto.so.4 => /usr/lib/x86_64-linux-gnu/libhcrypto.so.4 (0x00007f75a3852000)
        libroken.so.18 => /usr/lib/x86_64-linux-gnu/libroken.so.18 (0x00007f75a363c000)
        libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f75a3434000)
        libwind.so.0 => /usr/lib/x86_64-linux-gnu/libwind.so.0 (0x00007f75a320b000)
        libheimbase.so.1 => /usr/lib/x86_64-linux-gnu/libheimbase.so.1 (0x00007f75a2ffc000)
        libhx509.so.5 => /usr/lib/x86_64-linux-gnu/libhx509.so.5 (0x00007f75a2db2000)
        libsqlite3.so.0 => /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f75a2aa9000)
        libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f75a2871000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f75a24d3000)
```

**Comment by ethomson** [comment]
I think that `apt-get install libcurl4` should resolve the issue for now.  Many apologies, I'll get a prerelease of LibGit2Sharp that removes the curl dependency this weekend; I'll use a custom libgit2 branch if we haven't merged the libcurl removal by then.  😢 

Adding @bording in I'm mistaken, since he's the expert with the natives.

**Comment by ethomson** [comment]
Right - libcurl's packaging on Debian and Ubuntu it...  confusing.  Here's what I _think_ is happening:

You have the `libcurl3` package installed.  This includes `libcurl.so.4.5.0`.  The `libcurl4` package _also_ includes `libcurl.so.4.5.0`, but apparently they're not compatible?!  Despite having the same soversion?  This is...  frustrating.

@bording - Looking into this a bit, I _think_ that `libcurl4` is a drop-in replacement for `libcurl3`, but the opposite is not true.  I think that we may want to consider building and linking against `libcurl3` to be maximally compatible?  If we keep linking against `libcurl` at all, that is.

So anyway, yes, I think that if you run `apt-get install libcurl4` that it will remove `libcurl3` and that everything will start working.

**Comment by AnthonySteele** [comment]
Confirmed, `sudo apt-get install libcurl4` fixes the issue. 
I will update nukeeper docs.

**Comment by bording** [comment]
The changes you noticed in the latest preview were related to the fact that there is now a specific binary for Ubuntu 18.04 included, and it is built against `libcurl4`. The primary reason for this is that in a new 18.04 install, `libcurl4` is the "default" version installed. It's also the version that is installed in the Microsoft .NET Core docker image based on 18.04.

The goal is to minimize the amount of confusing, manual dependency installation that might need to happen, for as many of the distros supported by .NET Core as possible.

>@bording - Looking into this a bit, I think that libcurl4 is a drop-in replacement for libcurl3, but the opposite is not true. I think that we may want to consider building and linking against libcurl3 to be maximally compatible? If we keep linking against libcurl at all, that is.

@ethomson The generic `linux-x64` RID is built against `libcurl3` because that seems to be the only version that earlier vefsions of Ubuntu have, but 18.04 doesn't have `libcurl3` installed by default. As far as I can tell, neither of them are compatible with each other, which is the reason for things like the "/usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_3' not found" error mentioned in https://github.com/dotnet/roslyn/issues/2928.

By adding a binary for 18.04 in the latest preview, it appears that had the side effect of causing issues for people who had tried to work around the previous problem.

Ultimately, we just need to get rid of the libcurl dependency completely. In addition to `libcurl3` vs. `libcurl4`, there are also multiple different TLS backends that curl can use, and each distro has a different set of the libraries available/installed. This is how I feel at this point:

![wargames](https://user-images.githubusercontent.com/753669/48807117-856fd000-ecea-11e8-9436-8dbee6f0e472.jpg)

**Comment by AnthonySteele** [comment]
Right, thanks for all the help.

More in the spirit of post-incident analysis than anything else;  it seems to me that this was caused by the way that the specific documented setup (install `libcurl3` ) was no longer valid, and could and should be removed as the default setup worked; but I had no easy way of knowing this.  The short-term fix involves better docs?

The long term fix is to reduce dependencies in `libgit2sharp`. ironically, the whole NuKeeper project is about managing dependencies in .NET Code, and it gives you a stronger sense of how they have "weight". i.e. i.e. make it harder to move forward and so should be used with care not with abandon.

**Comment by ethomson** [comment]
> As far as I can tell, neither of them are compatible with each other, which is the reason for things like the "/usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_3' not found" error mentioned in dotnet/roslyn#2928.

Oy, how very strange.  I must be missing something, because this feels like the sort of thing that a new soversion would have solved.  Alas.

> The short-term fix involves better docs?

Yes; the short-term fix would be a documentation update for Ubuntu 18.04 users to select `libcurl4` instead of `libcurl3`.

> The long term fix is to reduce dependencies in libgit2sharp.

Agreed.  There are two pieces of work going on there in parallel, and I'd like to wrap this up this coming weekend.  Thanks for your patience and understanding.

**Comment by ethomson** [comment]
FYI, @AnthonySteele, I've released a prerelease LibGit2Sharp that should help a bit here.  I've updated libgit2 to have built-in support for proxies, so we no longer rely upon libcurl.  I think this should help the native dependency problems quite a bit.

This is available in [LibGit2Sharp 0.26.0-preview-0070](https://www.nuget.org/packages/LibGit2Sharp/0.26.0-preview-0070).
